File: LeetCode/278/278-First_Bad_Version.js
/**
 * Definition for isBadVersion()
 *
 * @param {integer} version number
 * @return {boolean} whether the version is bad
 * isBadVersion = function(version) {
 *     ...
 * };
 */

/**
 * @param {function} isBadVersion()
 * @return {function}
 */
var solution = function(isBadVersion) {
    /**
     * @param {integer} n Total versions
     * @return {integer} The first bad version
     */
    return function(n) {
      var s=0, e=n;
      while(e-s>1) {
        n = s+Math.floor((e-s)/2);
        if(isBadVersion(n)) e=n;
        else s=n;
      }
      return e;
    };
};

File: LeetCode/278/278-First_Bad_Version.cpp
// Forward declaration of isBadVersion API.
bool isBadVersion(int version);

class Solution {
public:
    int firstBadVersion(int n) {
      int a = 1, b = n;
      bool bn, bn1;
      n = a+(b-a)/2;
      while(a <= b) {
        bn = isBadVersion(n);
        bn1 = isBadVersion(n-1);
        if(bn && !bn1) break;
        if(bn)
          b = n - 1;
        else
          a = n + 1;
        n = a+(b-a)/2;
      }
      return n;
    }
};
